driver: lora: sx12xx: add method for channel activity detection

chihosin opened the pull request:
add ```lora_cad``` and ```lora_cad_async``` function to support channel activity detection (CAD) for Radio sx126x or llcc68, it has been verified on the SX1262 radio frequency chip that it can correctly detect that the channel is busy.

Signed-off-by: Chiho Sin [chihosin@icloud.com](mailto:chihosin@icloud.com)

martinjaeger reviewed (changes requested):
Thanks for your PR.

Can you please fix the compliance issues, which means remove the merge commits?

See also here for contribution workflow guidelines: https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow

Please don't open a new PR, but rewrite the commit history of this PR and use force-push to update it.

chihosin committed:
drivers: lora: sx12xx: add method for channel activity detection

add lora_cad and lora_cad_async function to support channel activity detection (CAD) for Radio sx126x or llcc68, it has been verified on the SX1262 radio frequency chip that it can correctly detect that the channel is busy.

Signed-off-by: Chiho Sin <chihosin@icloud.com>

martinjaeger commented on drivers/lora/sx12xx_common.c:
```diff
@@ -331,6 +376,98 @@ int sx12xx_lora_recv_async(const struct device *dev, lora_recv_cb cb)
+
+	/* Ensure available, decremented by sx12xx_ev_cad_done or on timeout */
+	if (dev_data.async_rx_cb) {
+		/* Increment atomic so both acquire and release will fail */
+		if (!atomic_cas(&dev_data.modem_usage, STATE_BUSY, STATE_CLEANUP)) {
+			return -EBUSY;
+		}
+		/* Put radio back into sleep mode */
+		Radio.Sleep();
+		/* Completely release modem */
+		dev_data.operation_done = NULL;
+		atomic_clear(&dev_data.modem_usage);
```
This looks exactly the same as `modem_release()`. Why was it copied here? Can't the function be called?

martinjaeger commented on drivers/lora/sx12xx_common.c:
```diff
@@ -331,6 +376,98 @@ int sx12xx_lora_recv_async(const struct device *dev, lora_recv_cb cb)
+	dev_data.operation_done = &done;
+	dev_data.cad_params.detected = &detected;
+
+#if defined(CONFIG_LORA_SX126X) || defined(CONFIG_LORA_STM32WL_SUBGHZ_RADIO) ||                    \
+	defined(CONFIG_LORA_STM32WL_SUBGHZ_RADIO)
```
Use `CONFIG_HAS_SEMTECH_SX126X` (see above).

martinjaeger commented on drivers/lora/sx12xx_common.c:
```diff
@@ -331,6 +376,98 @@ int sx12xx_lora_recv_async(const struct device *dev, lora_recv_cb cb)
+
+	/* Ensure available, decremented by sx12xx_ev_cad_done or on timeout */
+	if (dev_data.async_rx_cb) {
+		/* Increment atomic so both acquire and release will fail */
+		if (!atomic_cas(&dev_data.modem_usage, STATE_BUSY, STATE_CLEANUP)) {
+			return -EBUSY;
+		}
+		/* Put radio back into sleep mode */
+		Radio.Sleep();
+		/* Completely release modem */
+		dev_data.operation_done = NULL;
+		atomic_clear(&dev_data.modem_usage);
```
See above, use `modem_release`?
